agent/agent_pool.go

```diff
@@ -64,11 +93,81 @@ func (r *AgentPool) runWorker(ctx context.Context, worker *AgentWorker, im *Idle
 	defer worker.Disconnect(ctx)
 
 	// Starts the agent worker and wait for it to finish.
-	return worker.Start(ctx, im)
+	return worker.Start(ctx, r.idleMonitor)
 }
 
 func (r *AgentPool) Stop(graceful bool) {
 	for _, worker := range r.workers {
 		worker.Stop(graceful)
 	}
 }
+
+func agentWorkerHealthHandler(aw *AgentWorker) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		aw.stats.Lock()
+		defer aw.stats.Unlock()
+
+		if aw.stats.lastHeartbeatError != nil {
+			w.WriteHeader(http.StatusInternalServerError)
+			fmt.Fprintf(w, "ERROR: last heartbeat failed: %v. last successful was %v ago", aw.stats.lastHeartbeatError, time.Since(aw.stats.lastHeartbeat))
+		} else {
+			if aw.stats.lastHeartbeat.IsZero() {
+				fmt.Fprintf(w, "OK: no heartbeat yet")
+			} else {
+				fmt.Fprintf(w, "OK: last heartbeat successful %v ago", time.Since(aw.stats.lastHeartbeat))
+			}
+		}
+	}
+}
+
+func (ap *AgentPool) statusJSONHandler(l logger.Logger) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		type agentWorkerStatus struct {
+			Status       agentWorkerState `json:"status"`
+			CurrentJobID string           `json:"current_job_id,omitempty"`
+			ID           string           `json:"id"`
+			SpawnIndex   int              `json:"spawn_index"`
+		}
+
+		aggregateState := agentWorkerStateIdle
+		statuses := make([]agentWorkerStatus, 0, len(ap.workers))
+		for _, worker := range ap.workers {
+			// If any worker is busy, the aggregate state is busy
+			if worker.getState() == agentWorkerStateBusy {
+				aggregateState = agentWorkerStateBusy
+			}
+
+			statuses = append(statuses, agentWorkerStatus{
+				ID:           worker.agent.UUID,
+				Status:       worker.getState(),
+				CurrentJobID: worker.getCurrentJobID(),
+				SpawnIndex:   worker.spawnIndex,
+			})
+		}
+
+		err := json.NewEncoder(w).Encode(struct {
+			Health          string              `json:"health"`
+			AggregateStatus agentWorkerState    `json:"aggregate_status"`
+			WorkerStatuses  []agentWorkerStatus `json:"worker_statuses"`
```

Comment:
Could this just be `workers`? it's a little simpler, and the child elements are more than just status (also id, job id, etc)